### PR TITLE
statically render profile pages

### PR DESCRIPTION
**Pull request by KATT**
**⚡  Instant page load ~and instant navigation~ _(follow-up)_ ⚡** 

- Statically render "profile pages" + using tRPC to fetch data.

## Compare speeds of these two:

- https://cal.com/peer (~200-500ms on a **warm** start)
- https://calendso-git-feature-trpcuser-page-cal.vercel.app/peer (~20ms worldwide)

## What's actually going on?

- When visiting `/peer` we visit a **statically rendered** version of the page (which makes it instant)
- We use `getStaticProps` to pre-populate`react-query`'s cache of react-query so the data is there when we load the page
-  ~On mount of `/peer`~  Will be doing these as a follow-up - became harder since teams got added
   - ~We _prefetch_ the data of the route (in case we hit a stale cache you'll have a quick blimp)~
   - ~We _prefetch_ all the data of `/peer/[type]`-routes - making the navigation to `/peer/[type]` **instant**~

### CODE REVIEW

File: pages/[user].tsx [hunk 1]
```diff
@@ -6,45 +6,61 @@ import { ClockIcon, InformationCircleIcon, UserIcon } from "@heroicons/react/sol
...
+  const { isReady } = Theme(query.data?.user?.theme);
+  // prefetch all event types
+  useEffect(() => {
+    if (!query.data) {
+      return;
+    }
+    query.data.eventTypes.forEach(({ slug }) => {
+      utils.prefetchQuery(["booking.eventByUserAndType", { user: props.user, type: slug }]);
+    });
+  }, [props.user, query.data, utils]);
```
Comment (KATT) [inline pages/[user].tsx:26]: Makes sure navigation between `/peer` and `/peer/[type]` is instant by prefetching all events

File: pages/[user].tsx [hunk 2]
```diff
@@ -6,45 +6,61 @@ import { ClockIcon, InformationCircleIcon, UserIcon } from "@heroicons/react/sol
...
+  if (!query.data) {
+    // this only happens
+    // - in development
+    // - first request to a new user
+    // - 404 pages
+    return "...";
+  }
```
Comment (KATT) [inline pages/[user].tsx:33]: This actually shouldn't happen as we do `fallback: 'blocking'` which falls back on SSR

File: pages/[user].tsx [hunk 3]
```diff
@@ -89,51 +105,36 @@ export default function User(props: inferSSRProps<typeof getServerSideProps>) {
...
 }
-
-export const getServerSideProps = async (context: GetServerSidePropsContext) => {
-  const username = (context.query.user as string).toLowerCase();
-
-  const user = await prisma.user.findUnique({
-    where: {
-      username,
-    },
+export const getStaticPaths: GetStaticPaths = async () => {
```
Comment (KATT) [inline pages/[user].tsx:108]: This makes sure that we prerender all the users in the db. Will become heavy once we have tens of thousands and then there's other approaches to this as well.

File: pages/[user].tsx [hunk 4]
```diff
@@ -89,51 +105,36 @@ export default function User(props: inferSSRProps<typeof getServerSideProps>) {
...
+    paths: usernames.map((user) => ({
+      params: { user },
+    })),
+
+    // https://nextjs.org/docs/basic-features/data-fetching#fallback-blocking
+    fallback: "blocking",
+  };
+};
+
+export const getStaticProps = async (context: GetStaticPropsContext<{ user: string }>) => {
```
Comment (KATT) [inline pages/[user].tsx:125]: Prefetches the data on the page for each user and hydrates a cache that `react-query` uses on mount

**Review by KATT — COMMENTED** [review]
Some comments inline to guide reviewers

File: pages/[user].tsx (hunk 3 above)
Comment (mihaic195) [reply pages/[user].tsx:108]: What other approaches do you recommend?

Comment (KATT) [reply pages/[user].tsx:108]: What they refer to as "Incremental Static Regeneration". Essentially skipping `getStaticPaths` or reducing it to only build active users.

> For example, you can generate the most popular 1,000 products at build-time by returning the paths for the top 1,000 product IDs in getStaticPaths.

https://vercel.com/docs/next.js/incremental-static-regeneration?query=getstaticp#generating-paths

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -4,6 +4,19 @@ jobs:
...
+      BASE_URL: http://localhost:3000
+      JWT_SECRET: secret
+    services:
+      postgres:
+        image: postgres:12.1
+        env:
+          POSTGRES_USER: postgres
+          POSTGRES_DB: calendso
+        ports:
+          - 5432:5432
```
Comment (KATT) [inline .github/workflows/build.yml:19]: (build needs DB access now as we statically render pages)

File: pages/[user].tsx [hunk 6]
```diff
@@ -66,79 +76,43 @@ export default function User(props: inferSSRProps<typeof getServerSideProps>) {
...
-      name: true,
-      bio: true,
-      avatar: true,
-      theme: true,
-      plan: true,
+    },
+    where: {
+      // will statically render everyone on the PRO plan
+      // the rest will be statically rendered on first visit
+      plan: "PRO",
```
Comment (KATT) [inline pages/[user].tsx:87]: Made it so we only statically render paying customers

**Review by pumfleet — APPROVED** [review]

**Review by alishaz-polymath — APPROVED** [review]
Awesome stuff!!

**Pull request merged**